libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -298,9 +298,38 @@ void CompilerUtils::zeroInitialiseMemoryArray(ArrayType const& _type)
 	m_context << Instruction::SWAP1 << Instruction::POP;
 }
 
-void CompilerUtils::memoryCopy()
+void CompilerUtils::memoryCopy(bool _useIdentityPrecompile)
 {
 	// Stack here: size target source
+
+	if (!_useIdentityPrecompile)
+	{
+		m_context.appendInlineAssembly(R"(
+			{
+			// expects three locals: src, dst, len
+
+			// copy 32 bytes at once
+			start32:
+			jumpi(end32, lt(len, 32))
+			mstore(dst, mload(src))
+			dst := add(dst, 32)
+			src := add(src, 32)
+			len := sub(len, 32)
+			jump(start32)
+			end32:
+
+			// copy the last byte
```

Comment:
I think it should copy more than just a single byte. Also note that there are the `byte` and `mstore8` opcodes.
Perhaps something like this could work:
```
			let i := 0
			let remainder := mload(src)
		remainderLoop:
			mstore8(dst, byte(remainder, i))
			dst := add(dst, 1)
			i := add(i, 1)
			jumpi(remainderLoop, lt(i, len))
```
assuming that copies that are not multiples of 32 are quite rare.